### PR TITLE
feat(DatePicker): added ability to pass custom parsers for DatePicker's Input

### PR DESCRIPTION
**Pull request by sanex3339**
[FX-2324]

### Description

To implement human-readable date parsing for DatePicker in Staff-Portal we need to introduce a way to pass custom parsers for DatePicker's Input value. We should apply a custom parser on `onChange` event of DatePicker’s Input.

To have more details, please read all spike comments https://toptal-core.atlassian.net/browse/SPT-216

### How to test

- Open [Story Book](https://picasso.toptal.net/FX-2324-custom-date-picker-parsers/?path=/story/picasso-lab-datepicker--datepicker)
- Scroll down to the last example
- This DatePicker has a custom parser that parses the input value containing the year only (`2015`, `015`, `15`) and returns a random date within this year
- Selected day in the calendar should be updated when changing input value (`2015` => `2014`)
- Previous valid date should be selected when entering an invalid value (`abc`)

### Review

- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure that tests pass by running `yarn test`
- [x] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [x] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and
1. You have no extra requests.
2. You have optional requests.
   1. Add `nit: ` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because
1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:
1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-2324]: https://toptal-core.atlassian.net/browse/FX-2324?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx [hunk 1]
```diff
@@ -95,13 +94,14 @@ export interface Props
   weekStartsOn?: number
   /** IANA timezone to display and edit date(s) */
   timezone?: string
-  /* Invoked when input value has been changed. If method failed to parse a value, it must return undefined. Used to process input value before passing it to the `onChange` */
-  parseInputValue?: DatePickerStringParser
+  /** Custom parser for `DatePicker`'s input value to process custom input value, like, human-readable dates */
+  parseInputValue?: DatePickerInputCustomValueParser
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/DatePicker.tsx:98]: 👍

File: packages/picasso-lab/src/DatePicker/test.tsx [hunk 2]
```diff
@@ -253,95 +249,140 @@ describe('DatePicker', () => {
...
-      }
-
-      it('emits `string` if value is not a valid date', () => {
-        const handleChange = jest.fn()
-
-        const { getByPlaceholderText } = renderDatePicker({
-          ...defaultProps,
-          parseInputValue,
-          onChange: handleChange
+      describe('if parser returns `Date` value outside of the `min/max` dates', () => {
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/test.tsx:283]: I would use `describe` block to describe a condition

File: packages/picasso-lab/src/DatePicker/test.tsx [hunk 3]
```diff
@@ -253,95 +249,140 @@ describe('DatePicker', () => {
...
+        const parseInputValue: DatePickerInputCustomValueParser = jest
+          .fn()
+          .mockImplementation(() => {
+            return new Date(2021, 0, 1)
+          })
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/test.tsx:288]: why not?
```
jest.fn(() => new Date(2021, 0, 1))
```

**Review by dmaklygin1 — APPROVED** [review]

File: packages/picasso-lab/src/DatePicker/test.tsx (hunk 3 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/test.tsx:286]: https://github.com/toptal/picasso/pull/2298/commits/b052ad5d3ed73ddb8a3f1987349342c5ef47235b

File: packages/picasso-lab/src/DatePicker/test.tsx (hunk 2 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/test.tsx]: I'm not sure what did you mean. 
Anyway slightly changed texts
https://github.com/toptal/picasso/pull/2298/commits/b052ad5d3ed73ddb8a3f1987349342c5ef47235b

**Review by 13ANM — COMMENTED** [review]
LGTM but about the last testing point, no matter which value I save previously, when entering an invalid value like `abc` it takes me to `1902`, and `ab` to `202`

https://user-images.githubusercontent.com/79158829/144184275-e5c24e1e-bac5-4d2e-adc4-52055f741c5d.mov

**Comment by sanex3339** [comment]
@13ANM it's a bug in storybook example. Will fix it

**Comment by sanex3339** [comment]
@toptal-bot run package:alpha-release

**Review by TomasSlama — APPROVED** [review]
LGTM 👍

**Comment by sanex3339** [comment]
Please do not merge it until i test it

**Comment by sanex3339** [comment]
It was tested, everything is ok

**Comment by OleksandrNechai** [comment]
> Previous valid date should be selected when entering an invalid value (abc)

Peek 2021-12-01 20-22

:man_shrugging:

**Comment by sanex3339** [comment]
@OleksandrNechai You can reset the date

**Comment by OleksandrNechai** [comment]
I am sorry, my bad
Peek 2021-12-02 07-17
.

**Comment by sanex3339** [comment]
@OleksandrNechai so, yes. Empty date is a valid value, so after date reset if you enter invalid value like `abc` it will be reset to the last valid value, i.e. empty date. This behavior hasn't changed.

**Pull request merged**